File: 백준/DFS, BFS/1697.js
const fs = require("fs");

class Queue {
  constructor() {
    this.q = [];
    this.head = 0;
    this.rear = 0;
  }

  isEmpty() {
    return this.head === this.rear;
  }

  push(value) {
    this.q.push(value);
    this.rear++;
  }

  shift() {
    let rv = this.q[this.head];
    delete this.q[this.head++];
    return rv;
  }
}

const [s, k] = fs
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split(" ")
  .map((e) => Number(e));
const dx = [-1, 1, 2];
const checked = Array.from({ length: 100001 }, () => 0);
const q = new Queue();
q.push([s, 0]);

while (!q.isEmpty()) {
  const [cx, cnt] = q.shift();

  if (cx === k) {
    console.log(cnt);
    break;
  }

  dx.forEach((e) => {
    let nx;
    if (e === 2) {
      nx = cx * 2;
    } else {
      nx = cx + e;
    }

    if (nx >= 0 && nx <= 100000 && !checked[nx]) {
      checked[nx] = 1;
      q.push([nx, cnt + 1]);
    }
  });
}

File: 백준/DFS, BFS/1697.py
import sys
from collections import deque

n, k = map(int, sys.stdin.readline().split())

q = deque()
q.append([n, 0])
checked = [False] * 100001
checked[n] = True

while q:
    [cx, cd] = q.popleft()
    if cx == k:
        print(cd)
        break

    for nx in (cx - 1, cx + 1, cx * 2):
        if nx >= 0 and nx <= 100000 and not checked[nx]:
            checked[nx] = True
            q.append([nx, cd + 1])